File: app/src/main/java/io/dovid/multitimer/ui/MainActivity.java
package io.dovid.multitimer.ui;

import android.app.AlertDialog;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.os.Bundle;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SimpleItemAnimator;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import io.dovid.multitimer.BuildConfig;
import io.dovid.multitimer.R;
import io.dovid.multitimer.database.DatabaseHelper;
import io.dovid.multitimer.model.TimerDAO;
import io.dovid.multitimer.model.TimerEntity;
import io.dovid.multitimer.ui.preferences.PreferenceActivity;
import io.dovid.multitimer.utilities.AppRater;
import io.dovid.multitimer.utilities.RingtonePlayer;
import io.dovid.multitimer.utilities.VibrationPlayer;

public class MainActivity extends AppCompatActivity implements CreateTimerDialog.TimerCreateDialogListener, TimerUpdateDialog.TimerUpdateDialogListener {

    private RecyclerView recyclerView;
    private TimersAdapter timersAdapter;
    private FloatingActionButton fab;
    private int[] colors;

    private static final String TAG = "MAINACTIVITY";
    private static final int MATERIAL_THEME = 0;
    private static final int DARK_THEME = 1;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        // setup recycler view holding timers
        recyclerView = findViewById(R.id.recyclerView);
        // this avoid timer blinking animation when calling notifyItemChanged(position)
        ((SimpleItemAnimator) recyclerView.getItemAnimator()).setSupportsChangeAnimations(false);

        timersAdapter = new TimersAdapter(this);

        LinearLayoutManager manager = new LinearLayoutManager(this);
        if (recyclerView != null) {
            recyclerView.setLayoutManager(manager);
            recyclerView.setAdapter(timersAdapter);
        }

        setupColors();
        AppRater.app_launched(this);
        initSwap();

        fab = findViewById(R.id.fab);

        if (fab != null) {
            fab.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (timersAdapter != null && timersAdapter.getItemCount() < BuildConfig.MAX_TIMERS) {
                        CreateTimerDialog createDialog = CreateTimerDialog.getInstance();
                        createDialog.show(getFragmentManager(), "createTimer");
                    } else {
                        AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this).setTitle(R.string.main_max_number_of_timers);

                        builder.setMessage(getString(R.string.main_delete_before_adding_timer));
                        builder.setPositiveButton(getString(android.R.string.ok), new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                dialogInterface.dismiss();
                            }
                        }).create().show();
                    }
                }
            });
        }

        refreshTimers();
    }


    private void refreshTimers() {
        final Handler handler = new Handler();

        final int delay = 1000;

        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                timersAdapter.refreshTimers();
                handler.postDelayed(this, delay);
            }
        }, delay);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main_paid, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                Intent i = new Intent(this, PreferenceActivity.class);
                startActivity(i);
                break;
            case R.id.action_about:
                AboutDialog.getInstance().show(getFragmentManager(), "AboutDialog");
                break;
            case R.id.action_play_tutorial:
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    void setupColors() {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        int colorTheme = preferences.getInt(getString(R.string.preference_color_scheme), 0);

        if (colorTheme == MATERIAL_THEME) {
            if (fab != null) {
                fab.setBackgroundTintList(ColorStateList.valueOf(ContextCompat.getColor(this, R.color.colorPrimary)));
            }
            colors = ColorsThemePalette.getMaterialColors();
        } else if (colorTheme == DARK_THEME) {
            if (fab != null) {
                fab.setBackgroundTintList(ColorStateList.valueOf(ContextCompat.getColor(this, R.color.primary_dark_material_dark)));
            }
            colors = ColorsThemePalette.getDarkColors();
        } else {
            if (fab != null) {
                fab.setBackgroundTintList(ColorStateList.valueOf(ContextCompat.getColor(this, R.color.bakery_card2)));
            }

            colors = ColorsThemePalette.getBakeryColors();
        }
    }

    private void initSwap() {
        final ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new SwipeCallback(this) {
            @Override
            void swipeLeft(final int position) {
                timersAdapter.notifyDataSetChanged();
                new AlertDialog.Builder(MainActivity.this)
                        .setMessage("Are you sure you want this timer?")
                        .setNegativeButton("No", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                dialogInterface.dismiss();
                            }
                        })
                        .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                timersAdapter.deleteTimer(position);
                                timersAdapter.notifyItemRemoved(position);
                                timersAdapter.refreshTimers();
                                Log.d(TAG, "notifyItemRemoved at position: " + position);
                                dialogInterface.dismiss();
                            }
                        }).show();

            }

            @Override
            void swipeRight(int position) {
                if (timersAdapter != null) {
                    Log.d(TAG, "showUpdateTimerDialog");
                    timersAdapter.showUpdateTimerDialog(position);
                    timersAdapter.refreshTimers();
                    timersAdapter.notifyDataSetChanged();
                }
            }
        };

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(simpleItemTouchCallback);
        itemTouchHelper.attachToRecyclerView(recyclerView);
    }

    // TimerCreateDialogListener implementation
    public void onCreateTimer(String name, long time, DialogFragment dialog) {

        TimerDAO.create(DatabaseHelper.getInstance(this), name, time, false, true);
        if (timersAdapter != null) {
            timersAdapter.insertTimer(new TimerEntity(20, name, time, time, false, true));
        }

        dialog.dismiss();
    }

    // TimerUpdateDialogListener implementation
    @Override
    public void onUpdate(String name, long defaultTime, int timerId, DialogFragment dialog) {
        dialog.dismiss();
        TimerDAO.updateTimer(DatabaseHelper.getInstance(this), timerId, name, defaultTime, defaultTime);
        if (timersAdapter != null) {
            timersAdapter.refreshTimers();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        setupColors();
        if (timersAdapter != null) {
            timersAdapter.setColors(colors);
        }
        RingtonePlayer.stopPlaying();
        VibrationPlayer.stopVibrating();
    }

    @Override
    public void onBackPressed() {
        moveTaskToBack(true);
    }
}

File: app/src/main/java/io/dovid/multitimer/ui/MainActivity.kt
package io.dovid.multitimer.ui

import android.app.AlertDialog
import android.app.DialogFragment
import android.content.*
import android.content.res.ColorStateList
import android.os.Bundle
import android.os.Handler
import android.preference.PreferenceManager
import android.support.design.widget.FloatingActionButton
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import io.dovid.multitimer.BuildConfig
import io.dovid.multitimer.R
import io.dovid.multitimer.database.DatabaseHelper
import io.dovid.multitimer.model.TimerDAO
import io.dovid.multitimer.ui.preferences.PreferenceActivity
import io.dovid.multitimer.utilities.AppRater
import io.dovid.multitimer.utilities.RingtonePlayer
import io.dovid.multitimer.utilities.VibrationPlayer
import tourguide.tourguide.Overlay
import tourguide.tourguide.Pointer
import tourguide.tourguide.ToolTip
import tourguide.tourguide.TourGuide

class MainActivity : AppCompatActivity(), CreateTimerDialog.TimerCreateDialogListener, TimerUpdateDialog.TimerUpdateDialogListener {

    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: TimersAdapter? = null
    private var fab: FloatingActionButton? = null
    private var mTourGuideHandler: TourGuide? = null
    lateinit var colors: IntArray
    private lateinit var sharedPreferences: SharedPreferences

    private var toolbar: Toolbar? = null

    private var receiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            mAdapter?.refreshTimers()
            Log.d(TAG, "refreshing timers")
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this)

        // setup recycler view holding timers
        mRecyclerView = findViewById(R.id.recyclerView) as RecyclerView
        val manager = LinearLayoutManager(this)
        mRecyclerView?.layoutManager = manager
        mAdapter = TimersAdapter(this)
        mRecyclerView?.adapter = mAdapter

        setupColors()
        AppRater.app_launched(this)

        if (!sharedPreferences.contains(TutorialStep.POINT_TO_FAB.toString())) {
            loadTutorial(step = TutorialStep.POINT_TO_FAB, target = fab)
        }

        fab = findViewById(R.id.fab) as FloatingActionButton

        fab?.setOnClickListener {
            mTourGuideHandler?.cleanUp()
            sharedPreferences.edit().putBoolean("tutorial1", true).apply()

            if (mAdapter?.itemCount ?: 0 < BuildConfig.MAX_TIMERS) {
                val createDialog = CreateTimerDialog.getInstance()
                createDialog.show(fragmentManager, "createTimer")
            } else {
                val builder = AlertDialog.Builder(this@MainActivity).setTitle(getString(R.string.main_max_number_of_timers))

                builder.setMessage(getString(R.string.main_delete_before_adding_timer))

                builder.setPositiveButton(getString(android.R.string.ok), DialogInterface.OnClickListener { dialogInterface, _ ->
                    dialogInterface.dismiss()
                }).create().show()
            }
        }


    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_main_paid, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {

            R.id.action_settings -> {
                val i = Intent(this, PreferenceActivity::class.java)
                startActivity(i)
            }

            R.id.action_about -> {
                AboutDialog.getInstance().show(fragmentManager, "AboutDialog")
            }

            R.id.action_play_tutorial -> {
                sharedPreferences.edit()
                        .remove(TutorialStep.POINT_TO_FAB.toString())
                        .remove(TutorialStep.POINT_TO_TIMER.toString())
                        .commit()
                loadTutorial(step = TutorialStep.POINT_TO_FAB, target = fab)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    fun setupColors() {
        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val colorTheme = preferences.getString(getString(R.string.preference_color_scheme), "0")

        if (colorTheme == MATERIAL_THEME) {
            fab?.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(this, R.color.colorPrimary))
            colors = intArrayOf(R.color.material_card1,
                    R.color.material_card2,
                    R.color.material_card3,
                    R.color.material_card4,
                    R.color.material_card5,
                    R.color.material_card6,
                    R.color.material_card7,
                    R.color.material_card8,
                    R.color.material_card9,
                    R.color.material_card10)
        } else if (colorTheme == DARK_THEME) {
            fab?.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(this, R.color.primary_dark_material_dark))
            colors = intArrayOf(R.color.dark_card)
        } else {
            fab?.backgroundTintList = ColorStateList.valueOf(ContextCompat.getColor(this, R.color.bakery_card2))
            colors = intArrayOf(
                    R.color.bakery_card1,
                    R.color.bakery_card2,
                    R.color.bakery_card3,
                    R.color.bakery_card4,
                    R.color.bakery_card5)
        }
    }

    private fun loadTutorial(step: TutorialStep, target: View?) {
        if (target != null) {
            if (step.equals(TutorialStep.POINT_TO_FAB)) {
                mTourGuideHandler = TourGuide.init(this).with(TourGuide.Technique.Click)
                        .setPointer(Pointer().setGravity(Gravity.TOP))
                        .setToolTip(ToolTip().setTitle(getString(R.string.welcome)).
                                setDescription(getString(R.string.click_on_this_button)).setGravity(Gravity.TOP))
                        .setOverlay(Overlay().setOnClickListener {
                            mTourGuideHandler?.cleanUp()
                            sharedPreferences.edit()
                                    .putBoolean(TutorialStep.POINT_TO_FAB.toString(), true)
                                    .apply()
                        })
                        .playOn(target)
            } else if (step.equals(TutorialStep.POINT_TO_TIMER)) {
                mTourGuideHandler = TourGuide.init(this).with(TourGuide.Technique.VerticalDownward)
                        .setToolTip(ToolTip()
                                .setDescription(getString(R.string.long_press_timer)).setGravity(Gravity.CENTER).setOnClickListener {
                            mTourGuideHandler?.cleanUp()
                            sharedPreferences.edit()
                                    .putBoolean(TutorialStep.POINT_TO_TIMER.toString(), true)
                                    .apply()
                        })
                        .setOverlay(Overlay().setOnClickListener {
                            mTourGuideHandler?.cleanUp()
                        }.setStyle(Overlay.Style.Rectangle))
                        .playOn(target)
            } else {
                throw RuntimeException("Do not have a tutorial step number equals to " + step)
            }
        }
    }

    // TimerCreateDialogListener implementation
    override fun onCreateTimer(name: String, time: Long, dialog: DialogFragment) {

        TimerDAO.create(DatabaseHelper.getInstance(this), name, time, false, true)

        mAdapter?.insertTimer()
        dialog.dismiss()

        if (!sharedPreferences.contains(TutorialStep.POINT_TO_TIMER.toString())) {
            Handler().postDelayed({
                loadTutorial(step = TutorialStep.POINT_TO_TIMER, target = mRecyclerView?.getChildAt(0))
            }, 1000)
        }

        // in case user does not find easy to clear tutorial
        Handler().postDelayed({
            mTourGuideHandler?.cleanUp()
        }, 6000)

    }

    // TimerUpdateDialogListener implementation
    override fun onUpdate(name: String, defaultTime: Long, timerId: Int, dialog: DialogFragment) {
        dialog.dismiss()
        TimerDAO.updateTimer(DatabaseHelper.getInstance(this), timerId, name, defaultTime, defaultTime)
        mAdapter?.refreshTimers()
    }

    override fun onResume() {
        super.onResume()
        setupColors()
        mAdapter?.setColors(colors)
        registerReceiver(receiver, IntentFilter(BuildConfig.UPDATE_TIMERS))
        RingtonePlayer.stopPlaying()
        VibrationPlayer.stopVibrating()
    }

    override fun onStop() {
        super.onStop()
        try {
            unregisterReceiver(receiver)
        } catch (e: IllegalArgumentException) {
            Log.e(TAG, "MainActivity IllegalArgumentException", e);
        }
    }

    override fun onBackPressed() {
        moveTaskToBack(true)
    }

    companion object {
        private val TAG = "MAINACTIVITY"
        private val MATERIAL_THEME = "0"
        private val DARK_THEME = "1"
    }

    private enum class TutorialStep {
        POINT_TO_FAB, POINT_TO_TIMER
    }
}
